cloud-controller-manager/do/kubernetes_lb_validator_test.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2023 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/digitalocean/godo"
+
+	v1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrlruntimelog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+type annotations map[string]string
+
+type fakeRegionsService struct {
+	listFn func(context.Context, *godo.ListOptions) ([]godo.Region, *godo.Response, error)
+}
+
+func (f *fakeRegionsService) List(ctx context.Context, listOpts *godo.ListOptions) ([]godo.Region, *godo.Response, error) {
+	return f.listFn(ctx, listOpts)
+}
+
+func newFakeUnprocessableResponse() *godo.Response {
+	return newFakeResponse(http.StatusUnprocessableEntity)
+}
+
+func newFakeUnprocessableErrorResponse() *godo.ErrorResponse {
+	return &godo.ErrorResponse{
+		Response: &http.Response{
+			Request: &http.Request{
+				Method: "FAKE",
+				URL:    &url.URL{},
+			},
+			StatusCode: http.StatusUnprocessableEntity,
+			Body:       ioutil.NopCloser(bytes.NewBufferString("test")),
+		},
+	}
+}
+
+func Test_Handle(t *testing.T) {
+	os.Setenv(regionEnv, "nyc3")
+
+	testcases := []struct {
+		name            string
+		req             admission.Request
+		gCLient         *godo.Client
+		expectedAllowed bool
+		resp            *godo.Response
+		err             error
+		expectedMessage string
```

Comment:
I suggest we also add `expectedErrorStatusCode` that we populate for test cases where we expect `admission.Errored()` to be called. This allows us to assert that we pass down the upstream response code from the LB API response (that we fake out in our test) correctly to the caller.

As a bonus point, this also enables us to determine reliably if we should compare `expectedMessage` against the reason or the message field: it should be the former when `expectedErrorStatusCode` is zero and otherwise the latter.